### ISSUE TITLE
Calculate Mean first passage time

### ISSUE THREAD

**Issue by mitemijalkov** [softmatterlab/Braph-2.0-Matlab#137]
Calculate Mean first passage time. This can be implemented for all 4 types of graphs.

**Comment by maverickyuwei** [comment]
Dear @egolol @annacanal @mitemijalkov 
I have a question regarding this measure -- what if not always this measure could be computed?
I attached the corresponding function from BCT as follows.
As we can see within the codes for example, there might be two exceptions where this measure could not be computed
1. no eigenvalue of 1 found
2. no solution in the linear equations of adj matrix found

Have we encountered similar situation -- not always the measure could be computed -- before? How do you suggest to cope with it? return '0' or 'nan' instead?

Thank you for your time!

https://sites.google.com/site/bctnet/Home/functions/mean_first_passage_time.m?attredirects=0

>    aux = abs(diag(D_eigen)-1);
>     index = find(aux==min(aux));
>     if aux(index)>tol
>         error('cannot find eigenvalue of 1. Minimum eigenvalue value is %0.6f. Tolerance was set at %0.6f',aux(index)+1,tol);
>     end

**Comment by maverickyuwei** [comment]
Dear @giovannivolpe 
I am referring this question to you to hope to learn from your suggestions.

The equestion is: **What if a measure is not always calculable?**
Should the obj return a "0" or "null"?

The discussion about this question was quoted from previous comment.

"Mean first passage time" the measure from BCT would throw out an error (line 32-34) when the following conditions apply:
1.  no eigenvalue of 1 was found
2. the rank of input matrix was less than 2

https://sites.google.com/site/bctnet/Home/functions/mean_first_passage_time.m?attredirects=0

Thank you for your time reading this comment!

> Dear @egolol @annacanal @mitemijalkov
> I have a question regarding this measure -- what if not always this measure could be computed?
> I attached the corresponding function from BCT as follows.
> As we can see within the codes for example, there might be two exceptions where this measure could not be computed
> 
> 1. no eigenvalue of 1 found
> 2. no solution in the linear equations of adj matrix found
> 
> Have we encountered similar situation -- not always the measure could be computed -- before? How do you suggest to cope with it? return '0' or 'nan' instead?
> 
> Thank you for your time!
> 
> https://sites.google.com/site/bctnet/Home/functions/mean_first_passage_time.m?attredirects=0
> 
> > aux = abs(diag(D_eigen)-1);
> > index = find(aux==min(aux));
> > if aux(index)>tol
> > error('cannot find eigenvalue of 1. Minimum eigenvalue value is %0.6f. Tolerance was set at %0.6f',aux(index)+1,tol);
> > end

**Comment by annacanal** [comment]
> Dear @egolol @annacanal @mitemijalkov
> I have a question regarding this measure -- what if not always this measure could be computed?
> I attached the corresponding function from BCT as follows.
> As we can see within the codes for example, there might be two exceptions where this measure could not be computed
> 
> 1. no eigenvalue of 1 found
> 2. no solution in the linear equations of adj matrix found
> 
> Have we encountered similar situation -- not always the measure could be computed -- before? How do you suggest to cope with it? return '0' or 'nan' instead?
> 
> Thank you for your time!
> 
> https://sites.google.com/site/bctnet/Home/functions/mean_first_passage_time.m?attredirects=0
> 
> > aux = abs(diag(D_eigen)-1);
> > index = find(aux==min(aux));
> > if aux(index)>tol
> > error('cannot find eigenvalue of 1. Minimum eigenvalue value is %0.6f. Tolerance was set at %0.6f',aux(index)+1,tol);
> > end

I haven't encountered this situation before, but I would say to return '0'. Because the minimum mean first passage time cannot be zero when the calculation is possible, right?

**Comment by maverickyuwei** [comment]
Yes, @annacanal, it appears to me the same. To my best knowledge, this measure will take an average of (1 + m_kj),  since it is the number of steps from node i to j at random walker. (m_kj refers to any other transition node k apart from i or j)

**Comment by maverickyuwei** [comment]
Ergo, the  measure would never be zero. 
Thank you for your feedback Anna!

**Comment by mitemijalkov** [comment]
Hi @maverickyuwei @annacanal @egolol 

I think returning zero might be a good option for the reason you stated above. I think we should never return NaN because this has problems later with the statistical computations and plotting. However, if we return zero, I think we should write in the documentation why we do it, in order to avoid confusion (people might not expect zero from this measure so they might think it is a mistake).

One way to do this is to basically enforce the calculation only for the matrices for which it is possible. What I mean is, once the matrix is inputted for calculation, we could check whether the measure can be computed for this matrix and throw a warning/exception if it cannot with the explanation of why it cannot be computed.

This is something similar that we did in Braph 1 GUIs when the user was trying to input adjacency matrices that were not square, for example. It worked well enough in that case. However, I am not sure if we would like to do something similar in Braph 2.

**Comment by maverickyuwei** [comment]
Dear @mitemijalkov , @annacanal @egolol 

thank you for your feedback!
The warning/exception shall be necessary as I agree with you! And the example of making sure the input adj matrices remaining square is essential.

On this case, I will continue to head on this assignment with returning zero along with explanation comment in the code, and embed any further error/warning handling after we reach a conclusion on this matter.

Thank you again! And please feel free to let me know if I need to adjust.

**Comment by mitemijalkov** [comment]
That indeed seems like a good way to go. Eventually we will have to come up with a unified way to do this as there will be many measures that will have the same problem. But I guess, if needed, changing from zeros to NaN/Inf will be very easy :D

**Comment by maverickyuwei** [comment]
Will comply! thank you @mitemijalkov !

**Comment by egolol** [comment]
ERASE THIS
